policy/handler.go

```diff
@@ -199,49 +362,245 @@ func (h *Handler) updateHandler(updatedPolicy *sdk.ScalingPolicy) {
 	h.policyLock.Lock()
 	defer h.policyLock.Unlock()
 
+	h.log.Debug("updating check handlers", "old_checks", len(h.policy.Checks), "new_checks", len(updatedPolicy.Checks))
+
+	// Clear existing check handlers and load new ones.
+	h.checkRunners = nil
+
+	err := h.loadCheckRunners()
+	if err != nil {
+		h.errChn <- fmt.Errorf("unable to update policy, failed to load check handlers: %w", err)
+		return
+	}
+
+	h.log.Debug("check handlers updated", "count", len(h.checkRunners))
 	h.policy = updatedPolicy
 }
 
-// enforceCooldown blocks until the cooldown period has been reached, or the
-// handler has been instructed to exit. The boolean return details whether or
-// not the cooldown period passed without being interrupted.
-func (h *Handler) enforceCooldown(ctx context.Context, t time.Duration) bool {
+// applyMutators applies the mutators registered with the handler in order and
+// log any modification that was performed.
+func (h *Handler) applyMutators(p *sdk.ScalingPolicy) {
+	for _, m := range h.mutators {
+		for _, mutation := range m.MutatePolicy(p) {
+			h.log.Info("policy modified", "modification", mutation)
+		}
+	}
+}
+
+// Handle policy is the main part of the controller, it reads the target state,
+// gets the metrics and the necessary new count to keep up with the policy
+// and generates a scaling action if needed.
+func (h *Handler) CalculateNewCount(ctx context.Context, currentCount int64) (sdk.ScalingAction, error) {
+	h.log.Debug("received policy for evaluation")
+
+	// Record the start time of the eval portion of this function. The labels
+	// are also used across multiple metrics, so define them.
+	evalStartTime := nowFunc()
+	labels := []metrics.Label{
+		{Name: "policy_id", Value: h.policy.ID},
+		{Name: "target_name", Value: h.policy.Target.Name},
+	}
+
+	// Store check results by group so we can compare their results together.
+	checkGroups := make(map[string][]checkResult)
+
+	for _, ch := range h.checkRunners {
+		h.log.Debug("received policy check for evaluation")
+
+		metrics, err := ch.QueryMetrics(ctx)
+		if err != nil {
+			return sdk.ScalingAction{}, fmt.Errorf("failed to query source: %v", err)
+		}
+
+		action, err := ch.GetNewCountFromStrategy(ctx, currentCount, metrics)
+		if err != nil {
+			return sdk.ScalingAction{}, fmt.Errorf("failed get count from metrics: %v", err)
+
+		}
+
+		group := ch.check.Group
+		checkGroups[group] = append(checkGroups[group], checkResult{
+			action:  &action,
+			handler: ch,
+			group:   group,
+		})
+	}
+
+	winner := pickWinnerActionFromGroups(checkGroups)
+	if winner.handler == nil || winner.action == nil || winner.action.Direction == sdk.ScaleDirectionNone {
+		return sdk.ScalingAction{}, nil
+	}
+
+	h.log.Debug(fmt.Sprintf("check %s selected", winner.handler.check.Name),
+		"direction", winner.action.Direction, "count", winner.action.Count)
```

Comment:
This means we don't have to allocate for `fmt` until we actually log (ex. if you're logging at info) and lets you filter structured logs better:

```suggestion
	h.log.Debug("check selected", "name", winner.handler.check.Name,
		"direction", winner.action.Direction, "count", winner.action.Count)
```